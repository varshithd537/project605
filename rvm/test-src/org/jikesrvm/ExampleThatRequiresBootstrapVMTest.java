/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.jikesrvm;

import static org.junit.Assert.*;
import static org.jikesrvm.junit.runners.VMRequirements.isRunningOnBuiltJikesRVM;

import org.jikesrvm.junit.runners.VMRequirements;
import org.jikesrvm.junit.runners.RequiresBootstrapVM;
import org.jikesrvm.junit.runners.RequiresBuiltJikesRVM;
import org.junit.Test;
import org.junit.experimental.categories.Category;
import org.junit.runner.RunWith;

@RunWith(VMRequirements.class)
@Category(RequiresBootstrapVM.class)
public class ExampleThatRequiresBootstrapVMTest {

  @Test
  @Category(RequiresBuiltJikesRVM.class)
  public void testThatRequiresJikesRVM() {
    fail();
  }

  @Test
  @Category(RequiresBootstrapVM.class)
  public void testThatRequiresBootstrapVM() {
    assertFalse(isRunningOnBuiltJikesRVM());
  }

  @Test
  public void testWithNoRequirements() {
    assertFalse(isRunningOnBuiltJikesRVM());
  }
}
