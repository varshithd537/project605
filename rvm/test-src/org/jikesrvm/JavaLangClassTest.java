/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.jikesrvm;

import static org.junit.Assert.*;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.HashSet;

import org.jikesrvm.junit.runners.VMRequirements;
import org.junit.Test;
import org.junit.runner.RunWith;

/**
 * Jikes RVM replaces {@link java.lang.Class} from the class library with
 * its own version. This test is supposed to make sure that that version is
 * consistent with the official API.
 */
@RunWith(VMRequirements.class)
public class JavaLangClassTest {

  private static final Class<?>[] EMPTY_ARGS = new Class[0];

  @Test(expected = NullPointerException.class)
  public void forName_String_nullNameLeadsToNPE() throws ClassNotFoundException {
    Class.forName(null);
  }

  @Test(expected = NullPointerException.class)
  public void forName_MultipleArgs_nullNameLeadsToNPE() throws ClassNotFoundException {
    Class.forName(null, false, this.getClass().getClassLoader());
  }

  @Test(expected = NullPointerException.class)
  public void getMethod_nullNameLeadsToNPE() throws NoSuchMethodException {
    getClass().getMethod(null, EMPTY_ARGS);
  }

  @Test(expected = NoSuchMethodException.class)
  public void getMethod_initMethodCannotBeFound() throws NoSuchMethodException {
    getClass().getMethod("<init>", EMPTY_ARGS);
  }

  @Test(expected = NoSuchMethodException.class)
  public void getMethod_clinitMethodCannotBeFound() throws NoSuchMethodException {
    getClass().getMethod("<clinit>", EMPTY_ARGS);
  }

  @Test(expected = NullPointerException.class)
  public void getField_nullNameLeadsToNPE() throws NoSuchFieldException {
    getClass().getField(null);
  }

  @Test(expected = NullPointerException.class)
  public void getDeclaredField_nullNameLeadsToNPE() throws NoSuchFieldException {
    getClass().getDeclaredField(null);
  }

  @Test(expected = NullPointerException.class)
  public void getDeclaredMethod_nullNameLeadsToNPE() throws NoSuchMethodException {
    getClass().getDeclaredMethod(null);
  }

  @Test(expected = NullPointerException.class)
  public void getResourceAsStream_nullNameLeadsToNPE() throws NoSuchMethodException {
    getClass().getResourceAsStream(null);
  }

  @Test(expected = NullPointerException.class)
  public void getResource_nullArgumentLeadsToNPE() throws NoSuchMethodException {
    getClass().getResource(null);
  }

  @Test(expected = NullPointerException.class)
  public void getAnnotation_nullAnnotationLeadsToNPE() throws NoSuchMethodException {
    getClass().getAnnotation(null);
  }

  @Test(expected = NullPointerException.class)
  public void isAnnotationPresent_nullAnnotationLeadsToNPE() throws NoSuchMethodException {
    getClass().isAnnotationPresent(null);
  }

  @Test(expected = NullPointerException.class)
  public void isAssignalbleFrom_nullClassLeadsToNPE() throws NoSuchMethodException {
    getClass().isAssignableFrom(null);
  }

  @Test
  public void getMethodsForAbstractClassesDoesntReturnDuplicateImplementationmethods() throws NoSuchMethodException {
    Method[] methods = AbstractSubClass.class.getMethods();
    Method m = AbstractSubClass.class.getMethod("m");
    String methodName = m.getName();
    Class<?>[] methodParameters = m.getParameterTypes();
    HashSet<Method> foundMethods = new HashSet<Method>();
    for (Method method : methods) {
      if (methodName.equals(method.getName()) &&
      Arrays.equals(methodParameters, method.getParameterTypes())) {
        foundMethods.add(method);
      }
    }
    assertTrue(foundMethods.toString(), foundMethods.size() == 1);
  }

  private interface M1 {
    void m();
  }

  private interface M2 {
    void m();
  }

  private abstract static class AbstractSuperclass implements M2, M1 {
    @Override
    public void m() {
    }
  }

  private abstract static class AbstractSubClass extends AbstractSuperclass implements M2, M1 {
  }

  @Test
  public void getEnclosingMethodReturnsNullForLocalClassInType() throws NoSuchMethodException {
    class OuterLocalClass {
      class InnerLocalClass {}

      void getEnclosingMethodReturnsNullForLocalClassInType() {
        assertNull(InnerLocalClass.class.getEnclosingMethod());
      }
    }
    OuterLocalClass olc = new OuterLocalClass();
    olc.getEnclosingMethodReturnsNullForLocalClassInType();
  }

  private static class OuterClass {

    static {
      class InnerLocalClass {}
      assertNull(InnerLocalClass.class.getEnclosingMethod());
    }

    public void forceInitialization() { }
  }

  @Test
  public void getEnclosingMethodReturnsNullForLocalClassInStaticInitializer() throws NoSuchMethodException {
    OuterClass oc = new OuterClass();
    oc.forceInitialization();
  }

  @Test
  public void getEnclosingMethodReturnsNullForLocalClassInConstructor() throws NoSuchMethodException {
    class OuterLocalClass {
      OuterLocalClass() throws NoSuchMethodException {
        class LocalClassInConstructor {}
        assertNull(LocalClassInConstructor.class.getEnclosingMethod());
      }
    }
    @SuppressWarnings("unused")
    OuterLocalClass olc = new OuterLocalClass();
  }

  @Test
  public void getEnclosingMethodReturnsEnclosingMethodForLocalClassInMethod() throws NoSuchMethodException {
    class OuterLocalClass {
      public void methodWithLocalClass() throws NoSuchMethodException {
        class LocalClassInMethod {};
        assertEquals(LocalClassInMethod.class.getEnclosingMethod(),
            OuterLocalClass.class.getMethod("methodWithLocalClass"));
      }
    }
    OuterLocalClass olc = new OuterLocalClass();
    olc.methodWithLocalClass();
  }

  @Test
  public void getEnclosingConstructorReturnsNullForLocalClassInType() throws NoSuchMethodException {
    class OuterLocalClass {
      class InnerLocalClass {}

      void getEnclosingConstructorReturnsNullForLocalClassInType() {
        assertNull(InnerLocalClass.class.getEnclosingConstructor());
      }
    }
    OuterLocalClass olc = new OuterLocalClass();
    olc.getEnclosingConstructorReturnsNullForLocalClassInType();
  }

  private static class OuterClassForConstructor {

    static {
      class InnerLocalClass {}
      assertNull(InnerLocalClass.class.getEnclosingConstructor());
    }

    public void forceInitialization() {

    }
  }

  @Test
  public void getEnclosingConstructorReturnsNullForLocalClassInStaticInitializer() throws NoSuchMethodException {
    OuterClassForConstructor oc = new OuterClassForConstructor();
    oc.forceInitialization();
  }

  @Test
  public void getEnclosingConstructorReturnsNullForLocalClassInMethod() {
    class OuterLocalClass {
      void methodWithLocalClass() {
        class LocalClassInMethod {}
        assertNull(LocalClassInMethod.class.getEnclosingConstructor());
      }
    }
    OuterLocalClass olc = new OuterLocalClass();
    olc.methodWithLocalClass();
  }

  @Test
  public void getEnclosingMethodReturnsConstructorForLocalClassInConstructor() throws NoSuchMethodException {
    class OuterLocalClass {
      OuterLocalClass() throws NoSuchMethodException {
        class LocalClassInconstructor {}
        assertEquals(LocalClassInconstructor.class.getEnclosingConstructor(),
            OuterLocalClass.class.getDeclaredConstructor(org.jikesrvm.JavaLangClassTest.class));
      }
    }
    @SuppressWarnings("unused")
    OuterLocalClass olc = new OuterLocalClass();
  }

}
