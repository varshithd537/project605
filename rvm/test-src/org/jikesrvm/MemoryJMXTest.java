/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.jikesrvm;

import static org.hamcrest.number.OrderingComparison.greaterThanOrEqualTo;
import static org.junit.Assert.assertThat;

import java.lang.management.ManagementFactory;
import java.lang.management.MemoryMXBean;
import java.lang.management.MemoryUsage;
import org.jikesrvm.junit.runners.RequiresBuiltJikesRVM;
import org.jikesrvm.junit.runners.VMRequirements;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.experimental.categories.Category;
import org.junit.runner.RunWith;

@RunWith(VMRequirements.class)
@Category(RequiresBuiltJikesRVM.class)
public class MemoryJMXTest {

  private static MemoryMXBean memoryMXBean;

  @BeforeClass
  public static void setupMXBeans() {
    memoryMXBean = ManagementFactory.getMemoryMXBean();
  }

  @Test
  public void queryingMemoryMxBeanDoesntCauseErros() {
    MemoryUsage heapMemoryUsage = memoryMXBean.getHeapMemoryUsage();
    assertThat(heapMemoryUsage.getInit(), greaterThanOrEqualTo(-1L));
    MemoryUsage nonHeapMemoryUsage = memoryMXBean.getNonHeapMemoryUsage();
    assertThat(nonHeapMemoryUsage.getInit(), greaterThanOrEqualTo(-1L));
  }

  @Test
  public void objectPendingFinilizationCountIsNonNegative() throws Exception {
    assertThat(memoryMXBean.getObjectPendingFinalizationCount(), greaterThanOrEqualTo(0));
  }

}
