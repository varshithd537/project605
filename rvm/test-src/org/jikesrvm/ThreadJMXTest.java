/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.jikesrvm;

import static org.hamcrest.number.OrderingComparison.greaterThanOrEqualTo;
import static org.junit.Assert.assertThat;
import static org.hamcrest.CoreMatchers.*;

import java.lang.management.ManagementFactory;
import java.lang.management.ThreadInfo;
import java.lang.management.ThreadMXBean;

import org.jikesrvm.junit.runners.RequiresBuiltJikesRVM;
import org.jikesrvm.junit.runners.VMRequirements;
import org.jikesrvm.scheduler.SystemThread;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.experimental.categories.Category;
import org.junit.runner.RunWith;
import org.vmmagic.pragma.NonMoving;

@RunWith(VMRequirements.class)
@Category(RequiresBuiltJikesRVM.class)
public class ThreadJMXTest {

  private static ThreadMXBean threadJmxBean;

  @BeforeClass
  public static void setupMXBeans() {
    threadJmxBean = ManagementFactory.getThreadMXBean();
  }

  @Test
  public void queryingThreadMxBeanDoesntCauseErros() {
    long tid = Thread.currentThread().getId();
    ThreadInfo tInfo = threadJmxBean.getThreadInfo(tid);
    assertThat(tInfo, notNullValue());
  }

  @Test
  public void queryingThreadMxBeanForSystemThreadYieldsNull() {
    SystemThread st = new TestSystemThread();
    st.start();
    long tid = st.getRVMThread().getJavaLangThread().getId();
    ThreadInfo tInfo = threadJmxBean.getThreadInfo(tid);
    assertThat(tInfo, nullValue());
  }


  @Test
  public void queryingInvalidThreadIdsYieldsNull() {
    long tid = Long.MAX_VALUE / 2;
    ThreadInfo tInfo = threadJmxBean.getThreadInfo(tid);
    assertThat(tInfo, nullValue());
  }


  @Test
  public void arrayOfThreadIdsIsNonEmpty() throws Exception {
    assertThat(threadJmxBean.getAllThreadIds().length, greaterThanOrEqualTo(0));
  }

  @NonMoving
  private static class TestSystemThread extends SystemThread {

    TestSystemThread() {
      super("test-system-thread");
    }

    @Override
    public void run() {
      while (true) {
        try {
          synchronized (this) {
            this.wait();
          }
        } catch (IllegalMonitorStateException e) {
          throw new RuntimeException(e);
        } catch (InterruptedException e) {
          throw new RuntimeException(e);
        }
      }
    }
  }

}
