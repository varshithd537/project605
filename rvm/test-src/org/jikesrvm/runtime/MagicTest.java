/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.jikesrvm.runtime;

import static org.junit.Assert.*;
import static org.hamcrest.CoreMatchers.*;
import static org.jikesrvm.compilers.common.CompiledMethod.*;
import org.jikesrvm.classloader.NormalMethod;
import org.jikesrvm.compilers.common.CompiledMethod;
import org.jikesrvm.compilers.common.RuntimeCompiler;
import org.jikesrvm.compilers.opt.runtimesupport.OptCompiledMethod;
import org.jikesrvm.junit.runners.RequiresBuiltJikesRVM;
import org.jikesrvm.junit.runners.RequiresOptCompiler;
import org.jikesrvm.junit.runners.VMRequirements;
import org.jikesrvm.tests.util.TestingTools;
import org.junit.After;
import org.junit.Test;
import org.junit.experimental.categories.Category;
import org.junit.runner.RunWith;

/**
 * Test cases for {@link Magic} that either need high-level operations to test or aren't required for
 * correct operation of the VM. Most tests for methods from Magic should be written for TestMagic from
 * the basic test run. In contrast to unit tests run on the built RVM image, the basic test run doesn't
 * assume a mostly working VM.
 */
@RunWith(VMRequirements.class)
@Category(RequiresBuiltJikesRVM.class)
public class MagicTest {

  private NormalMethod testMethodForGetCompilerLevel;

  @After
  public void invalidateCompiledMethods() {
    testMethodForGetCompilerLevel.invalidateCompiledMethod(testMethodForGetCompilerLevel.getCurrentCompiledMethod());
  }

  @Test
  public void getCompilerLevelMapsToTheCorrectConstantsForBaseline() throws Exception {
    testMethodForGetCompilerLevel = TestingTools.getNormalMethod(MagicTest.class, "getCompilerLevel");
    CompiledMethod cm = RuntimeCompiler.baselineCompile(testMethodForGetCompilerLevel);
    assertThat(cm.getCompilerType(), is(BASELINE));
    assertThat(getCompilerLevel(), is(-1));
  }

  public void assertThatGetCompilerLevelisCorrectForOptLevel(int optLevel) {
    recompileTestMethodOnOptLevel(optLevel);
    assertThat(getCompilerLevel(), is(optLevel));
  }

  public void recompileTestMethodOnOptLevel(int optLevel) {
    int methodId = RuntimeCompiler.recompileWithOpt(testMethodForGetCompilerLevel, optLevel);
    assertThat(methodId, not(equalTo(-1)));
    CompiledMethod cm = testMethodForGetCompilerLevel.getCurrentCompiledMethod();
    assertThat(cm.getCompilerType(), is(OPT));
    assertThat(((OptCompiledMethod)cm).getOptLevel(), is(optLevel));
  }

  @Test
  @Category(RequiresOptCompiler.class)
  public void getCompilerLevelMapsToTheCorrectConstantsForOpt() throws Exception {
    testMethodForGetCompilerLevel = TestingTools.getNormalMethod(MagicTest.class, "getCompilerLevel");
    assertThatGetCompilerLevelisCorrectForOptLevel(0);
    assertThatGetCompilerLevelisCorrectForOptLevel(1);
    assertThatGetCompilerLevelisCorrectForOptLevel(2);
  }

  public static int getCompilerLevel() {
    return Magic.getCompilerLevel();
  }

}
