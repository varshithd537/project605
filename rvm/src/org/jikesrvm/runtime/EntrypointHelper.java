/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.jikesrvm.runtime;

import org.jikesrvm.VM;
import org.jikesrvm.classloader.Atom;
import org.jikesrvm.classloader.BootstrapClassLoader;
import org.jikesrvm.classloader.RVMClass;
import org.jikesrvm.classloader.RVMField;
import org.jikesrvm.classloader.RVMMember;
import org.jikesrvm.classloader.RVMMethod;
import org.jikesrvm.classloader.MethodReference;
import org.jikesrvm.classloader.NormalMethod;
import org.jikesrvm.classloader.TypeReference;
import org.vmmagic.pragma.Entrypoint;

/**
 * Helper class for retrieving entrypoints. Entrypoints are fields and
 * methods of the virtual machine that are needed by compiler-generated
 * machine code or C runtime code.
 */
public class EntrypointHelper {
  /**
   * Get description of virtual machine component (field or method).
   * <p>
   * Note: This is method is intended for use only by VM classes that need
   * to address their own fields and methods in the runtime virtual machine
   * image.  It should not be used for general purpose class loading.
   * @param classDescriptor  class  descriptor - something like "Lorg/jikesrvm/RuntimeEntrypoints;"
   * @param memberName       member name       - something like "invokestatic"
   * @param memberDescriptor member descriptor - something like "()V"
   * @return corresponding RVMMember object
   */
  private static RVMMember getMember(String classDescriptor, String memberName, String memberDescriptor) {
    Atom clsDescriptor = Atom.findOrCreateAsciiAtom(classDescriptor);
    Atom memName = Atom.findOrCreateAsciiAtom(memberName);
    Atom memDescriptor = Atom.findOrCreateAsciiAtom(memberDescriptor);
    try {
      TypeReference tRef =
          TypeReference.findOrCreate(BootstrapClassLoader.getBootstrapClassLoader(), clsDescriptor);
      RVMClass cls = (RVMClass) tRef.resolve();
      cls.resolve();

      RVMMember member;
      if ((member = cls.findDeclaredField(memName, memDescriptor)) != null) {
        verifyThatFieldIsNotFinal((RVMField) member);
        verifyPresenceOfEntrypointAnnotation(member);
        return member;
      }
      if ((member = cls.findDeclaredMethod(memName, memDescriptor)) != null) {
        verifyPresenceOfEntrypointAnnotation(member);
        return member;
      }
    } catch (Exception e) {
      e.printStackTrace();
    }
    // The usual causes for getMember() to fail are:
    //  1. you mispelled the class name, member name, or member signature
    //  2. the class containing the specified member didn't get compiled
    //
    VM.sysWriteln("Entrypoints.getMember: can't resolve class=" +
                classDescriptor +
                " member=" +
                memberName +
                " desc=" +
                memberDescriptor);
    if (VM.VerifyAssertions) VM._assert(VM.NOT_REACHED);
    return null;
  }

  private static void verifyThatFieldIsNotFinal(RVMField field) {
    if (field.isFinal() && field.isAnnotationPresent(Entrypoint.class) &&
        !field.getAnnotation(Entrypoint.class).fieldMayBeFinal()) {
      String msg = "ERROR: Field " + field +
          " is marked with @Entrypoint annotation and is final." +
          "This is forbidden because the Java compiler views a final field " +
          "as immutable which it likely won't be if it's " +
          "explicitly accessed by code. If it's indeed final, use " +
          "@Entrypoint(fieldMayBeFinal = true) to mark the field.";
      throw new Error(msg);
    }
  }

  private static void verifyPresenceOfEntrypointAnnotation(RVMMember member) {
    if (VM.VerifyAssertions && !(member.isAnnotationPresent(Entrypoint.class))) {
      // For certain methods, it's clear that they're accessed by the
      // compiler, so an annotation is not required:
      boolean annotationRequired = true;
      if (member instanceof RVMMethod) {
        RVMMethod m = (RVMMethod) member;
        RVMClass declClass = m.getDeclaringClass();
        if (declClass.getTypeRef().isMagicType() ||
            declClass.getTypeRef().isUnboxedType() ||
            // don't impose constraints on class library methods,
            // it's the VM's job to handle those correctly
            declClass.getPackageName().startsWith("java.lang")) {
          annotationRequired = false;
        }
      }
      // Don't require annotations on fields for the class library.
      if (member instanceof RVMField) {
        RVMField field = (RVMField) member;
        RVMClass declClass = field.getDeclaringClass();
        if (declClass.getPackageName().startsWith("java.lang")) {
          annotationRequired = false;
        }
      }

      if (annotationRequired) {
        String msg = "WARNING: MISSING @Entrypoint ANNOTATION: " + member +
            " is missing an @Entrypoint annotation!";
          throw new Error(msg);
      }
    }
  }

  public static NormalMethod getMethod(String klass, String member, String descriptor, final boolean runtimeServiceMethod) {
    NormalMethod m = (NormalMethod) getMember(klass, member, descriptor);
    m.setRuntimeServiceMethod(runtimeServiceMethod);
    return m;
  }

  public static NormalMethod getMethod(String klass, String member, String descriptor) {
    return getMethod(klass, member, descriptor, true);
  }

  private static String makeDescriptor(Class<?>... argTypes) {
    Class<?> lastClass = null;
    StringBuilder result = new StringBuilder("(");
    for (Class<?> c: argTypes) {
      if (lastClass != null) {
        result.append(TypeReference.findOrCreate(lastClass).getName().toString());
      }
      lastClass = c;
    }
    result.append(")").append(TypeReference.findOrCreate(lastClass).getName().toString());
    return result.toString();
  }

  public static RVMMethod getMethod(Class<?> klass, Atom member, Class<?>... argTypes) {
    if (!VM.runningVM) { // avoid compiling this code into the boot image
      try {
        TypeReference tRef = TypeReference.findOrCreate(klass);
        RVMClass cls = tRef.resolve().asClass();
        cls.resolve();

        Atom descriptor = Atom.findOrCreateAsciiAtom(makeDescriptor(argTypes));

        RVMMethod method = cls.findDeclaredMethod(member, descriptor);
        if (method != null) {
          verifyPresenceOfEntrypointAnnotation(method);
          return method;
        }
      } catch (Throwable t) {
        throw new Error("Entrypoints.getMethod: can't resolve class=" +
            klass + " member=" + member + " desc=" + makeDescriptor(argTypes), t);
      }
    }
    throw new Error("Entrypoints.getMethod: can't resolve class=" +
        klass + " member=" + member + " desc=" + makeDescriptor(argTypes));
  }

  public static MethodReference getMethodReference(Class<?> klass, Atom member, Class<?>... argTypes) {
    if (!VM.runningVM) { // avoid compiling this code into the boot image
      TypeReference tRef = TypeReference.findOrCreate(klass);
      if (tRef.resolve().isClassType()) {
        return getMethod(klass, member, argTypes).getMemberRef().asMethodReference();
      } else { // handle method references to unboxed types
        Atom descriptor = Atom.findOrCreateAsciiAtom(makeDescriptor(argTypes));
        return MethodReference.findOrCreate(tRef, member, descriptor);
      }
    }
    throw new Error("Entrypoints.getMethod: can't resolve class=" +
        klass + " member=" + member + " desc=" + makeDescriptor(argTypes));
  }

  public static RVMField getField(String klass, String member, String descriptor) {
    return (RVMField) getMember(klass, member, descriptor);
  }

  /**
   * Get description of virtual machine field.
   * @param klass class containing field
   * @param member member name - something like "invokestatic"
   * @param type of field
   * @return corresponding RVMField
   */
  public static RVMField getField(Class<?> klass, String member, Class<?> type) {
    if (!VM.runningVM) { // avoid compiling this code into the boot image
      try {
        TypeReference klassTRef = TypeReference.findOrCreate(klass);
        RVMClass cls = klassTRef.resolve().asClass();
        cls.resolve();

        Atom memName = Atom.findOrCreateAsciiAtom(member);
        Atom typeName = TypeReference.findOrCreate(type).getName();

        RVMField field = cls.findDeclaredField(memName, typeName);
        if (field != null) {
          verifyPresenceOfEntrypointAnnotation(field);
          verifyThatFieldIsNotFinal(field);
          return field;
        }
      } catch (Throwable t) {
        throw new Error("Entrypoints.getField: can't resolve class=" +
            klass + " member=" + member + " desc=" + type, t);
      }
    }
    throw new Error("Entrypoints.getField: can't resolve class=" +
        klass + " member=" + member + " desc=" + type);
  }

  /**
   * Get description of virtual machine field.
   * @param klass class containing field
   * @param member member name - something like "invokestatic"
   * @param type of field
   * @return corresponding RVMField
   */
  static RVMField getField(String klass, String member, Class<?> type) {
    if (!VM.runningVM) { // avoid compiling this code into the boot image
      try {
        TypeReference tRef = TypeReference.findOrCreate(klass);
        RVMClass cls = tRef.resolve().asClass();
        cls.resolve();

        Atom memName = Atom.findOrCreateAsciiAtom(member);
        Atom typeName = TypeReference.findOrCreate(type).getName();

        RVMField field = cls.findDeclaredField(memName, typeName);
        if (field != null) {
          verifyPresenceOfEntrypointAnnotation(field);
          verifyThatFieldIsNotFinal(field);
          return field;
        }
      } catch (Throwable t) {
        throw new Error("Entrypoints.getField: can't resolve class=" +
            klass + " member=" + member + " desc=" + type, t);
      }
    }
    throw new Error("Entrypoints.getField: can't resolve class=" +
        klass + " member=" + member + " desc=" + type);
  }

  /**
   * Get description of virtual machine method.
   * @param klass class  containing method
   * @param member member name - something like "invokestatic"
   * @param descriptor member descriptor - something like "()V"
   * @return corresponding RVMMethod
   */
  public static NormalMethod getMethod(Class<?> klass, String member, String descriptor) {
    if (!VM.runningVM) { // avoid compiling this code into the boot image
      try {
        TypeReference klassTRef = TypeReference.findOrCreate(klass);
        RVMClass cls = klassTRef.resolve().asClass();
        cls.resolve();

        Atom memName = Atom.findOrCreateAsciiAtom(member);
        Atom memDescriptor = Atom.findOrCreateAsciiAtom(descriptor);

        NormalMethod m = (NormalMethod)cls.findDeclaredMethod(memName, memDescriptor);
        if (m != null) {
          verifyPresenceOfEntrypointAnnotation(m);
          m.setRuntimeServiceMethod(true);
          return m;
        }
      } catch (Throwable t) {
        throw new Error("Entrypoints.getField: can't resolve class=" +
            klass + " member=" + member + " desc=" + descriptor, t);
      }
    }
    throw new Error("Entrypoints.getMethod: can't resolve class=" +
        klass + " method=" + member + " desc=" + descriptor);
  }
}
