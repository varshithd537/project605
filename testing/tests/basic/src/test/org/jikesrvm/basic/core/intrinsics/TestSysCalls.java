/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package test.org.jikesrvm.basic.core.intrinsics;

import org.jikesrvm.VM;
import org.jikesrvm.runtime.SysCall;

public class TestSysCalls {

  public static void main(String[] args) {
    testArgumentPassingForSysCalls();
  }

  public static void testArgumentPassingForSysCalls() {
    VM.sysWriteln("Test argument passing: longs and doubles");

    SysCall.sysCall.sysArgumentPassingSeveralLongsAndSeveralDoubles(Long.MIN_VALUE, Long.MAX_VALUE, -1L, 1L, 0xFFFF0000FFFF0000L, 0x0000FFFF0000FFFFL, 0, -1L, Double.MIN_VALUE, Double.MAX_VALUE, -1D, 0.1D, 1.2345678910E300D, Double.NaN, Double.MAX_VALUE, -1D);

    VM.sysWriteln("Test argument passing: floats and ints");

    SysCall.sysCall.sysArgumentPassingSeveralFloatsAndSeveralInts(Float.MAX_VALUE, Float.MIN_VALUE, -1f, 1f, 0.1f, 1.23456789E34F, Float.NaN, Float.POSITIVE_INFINITY, Integer.MAX_VALUE, Integer.MIN_VALUE, -1, 1, 0, 0xFFFF0000, 0x0000FFFF, 0xF0E0A1B1);
  }

}
