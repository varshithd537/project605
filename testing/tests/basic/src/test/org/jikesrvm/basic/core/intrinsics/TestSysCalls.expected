Test argument passing: longs and doubles
First long -9223372036854775808
Second long 9223372036854775807
Third long -1
Fourth long 1
Fifth long -281470681808896
Sixth long 281470681808895
Seventh long 0
Eight long -1
First double 0.000000
Second double 179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368.000000
Third double -1.000000
Fourth double 0.100000
Fifth double 1234567890999999930732114707826946781561496553158231368735739722485522148589380095997427052998754393345483556542327289145867568434814350380373152620607527753392384540950953066024268311075889205725759879376963497031443323726116915128397656892239086442753924758455253355700751236401954380697235898761216.000000
Sixth double nan
Seventh double 179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368.000000
Eight double -1.000000
Test argument passing: floats and ints
First float 340282346638528859811704183484516925440.000000
Second float 0.000000
Third float -1.000000
Fourth float 1.000000
Fifth float 0.100000
Sixth float 12345678906183668990302990096138240.000000
Seventh float nan
Eight float inf
First int 2147483647
Second int -2147483648
Third int -1
Fourth int 1
Fifth int 0
Sixth int -65536
Seventh int 65535
Eight int -253713999
